<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in New York City</font></h1>

## Introduction

In this lab, you will learn how to convert addresses into their equivalent latitude and longitude values. Also, you will use the Foursquare API to explore neighborhoods in New York City. You will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. You will use the *k*-means clustering algorithm to complete this task. Finally, you will use the Folium library to visualize the neighborhoods in New York City and their emerging clusters.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in New York City</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

## 1. Download and Explore Dataset

Neighborhood has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood. 

Luckily, this dataset exists for free on the web. Feel free to try to find this dataset on your own, but here is the link to the dataset: https://geo.nyu.edu/catalog/nyu_2451_34572

For your convenience, I downloaded the files and placed it on the server, so you can simply run a `wget` command and access the data. So let's go ahead and do that.

In [2]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [70]:
from bs4 import BeautifulSoup

#### Load and explore the data

Next, let's load the data.

In [2]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Let's take a quick look at the data.

In [3]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

# Json Data
{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
 

Notice how all the relevant data is in the *features* key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [4]:
neighborhoods_data = newyork_data['features']

Let's take a look at the first item in this list.

In [7]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the data into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a *pandas* dataframe. So let's start by creating an empty dataframe.

In [5]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [6]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [11]:
for data in neighborhoods_data:
    borough = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [8]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [9]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [12]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


#### Create a map of New York with neighborhoods superimposed on top.

In [13]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [14]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [15]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7900869, -73.9598295.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [19]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [20]:
CLIENT_ID = '24I53K5BPDOO0SPEUWIPDMI1GA3BM5RN2KL2EXY3YPRNSTEW' # your Foursquare ID
CLIENT_SECRET = 'MXSNHIAQHEZNEVQ02POGGWSSG15DXM2K3XIMHCX5PLQYOOON' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 24I53K5BPDOO0SPEUWIPDMI1GA3BM5RN2KL2EXY3YPRNSTEW
CLIENT_SECRET:MXSNHIAQHEZNEVQ02POGGWSSG15DXM2K3XIMHCX5PLQYOOON


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [21]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [22]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [23]:
# type your answer here
radius=500
LIMIT=100
url='https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&query={}&radius={}&limit={}'.format(CLIENT_ID,CLIENT_SECRET,VERSION,neighborhood_latitude,neighborhood_longitude,neighborhood_name,radius,LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=24I53K5BPDOO0SPEUWIPDMI1GA3BM5RN2KL2EXY3YPRNSTEW&client_secret=MXSNHIAQHEZNEVQ02POGGWSSG15DXM2K3XIMHCX5PLQYOOON&v=20180605&ll=40.87655077879964,-73.91065965862981&query=Marble Hill&radius=500&limit=100'

Double-click __here__ for the solution.
<!-- The correct answer is:
LIMIT = 100 # limit of number of venues returned by Foursquare API
-->

<!--
radius = 500 # define radius
-->

<!--
\\ # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
--> 

Send the GET request and examine the resutls

In [24]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cf3d874dd57970e257d9a53'},
 'response': {'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'query': 'marble hill',
  'totalResults': 7,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b5e0e1ff964a520357b29e3',
       'name': 'Metro North - Marble Hill Train Station',
       'location': {'address': '1 West 225th St',
        'lat': 40.87469809860372,
        'lng': -73.91205168105456,
        'labeledLatLngs': [{'label': 'display',
          'lat': 40.87469809860372,
          'lng': -73.91205168105456}],

# Sample Expore JSON data
{'meta': {'code': 200, 'requestId': '5cf3d874dd57970e257d9a53'},
 'response': {'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'query': 'marble hill',
  'totalResults': 7,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b5e0e1ff964a520357b29e3',
       'name': 'Metro North - Marble Hill Train Station',
       'location': {'address': '1 West 225th St',
        'lat': 40.87469809860372,
        'lng': -73.91205168105456,
        'labeledLatLngs': [{'label': 'display',
          'lat': 40.87469809860372,
          'lng': -73.91205168105456}],
        'distance': 237,
        'postalCode': '10463',
        'cc': 'US',
        'city': 'New York',
        'state': 'NY',
        'country': 'United States',
        'formattedAddress': ['1 West 225th St',
         'New York, NY 10463',
         'United States']},
       'categories': [{'id': '4bf58dd8d48988d129951735',
         'name': 'Train Station',
         'pluralName': 'Train Stations',
         'shortName': 'Train Station',
         'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/trainstation_',
          'suffix': '.png'},
         'primary': True}],
       'photos': {'count': 0, 'groups': []}},
      'referralId': 'e-0-4b5e0e1ff964a520357b29e3-0'}

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [25]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [27]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON
print(type(nearby_venues))
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<class 'pandas.core.frame.DataFrame'>


name      categories        lat  \
0  Metro North - Marble Hill Train Station   Train Station  40.874698   
1                  Marble Hill Youthmarket  Farmers Market  40.874519   
2                    marble  hill pharmacy        Pharmacy  40.875050   
3    MTA Subway - 225th St/Marble Hill (1)   Metro Station  40.874486   
4                   Marble Hill Playground      Playground  40.877765   

         lng  
0 -73.912052  
1 -73.910394  
2 -73.909195  
3 -73.909589  
4 -73.907994

And how many venues were returned by Foursquare?

In [29]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))
nearby_venues

7 venues were returned by Foursquare.


name      categories  \
0            Metro North - Marble Hill Train Station   Train Station   
1                            Marble Hill Youthmarket  Farmers Market   
2                              marble  hill pharmacy        Pharmacy   
3              MTA Subway - 225th St/Marble Hill (1)   Metro Station   
4                             Marble Hill Playground      Playground   
5                Orange Park, Marble Hill, Bronx, NY            Park   
6  MTA MaBSTOA Bus at W. 225th Street & Broadway ...     Bus Station   

         lat        lng  
0  40.874698 -73.912052  
1  40.874519 -73.910394  
2  40.875050 -73.909195  
3  40.874486 -73.909589  
4  40.877765 -73.907994  
5  40.877986 -73.908028  
6  40.874318 -73.909297

<a id='item2'></a>

## 2. Explore Neighborhoods in Manhattan

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [53]:
neighborhoods_data = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )


Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [55]:
neighborhoods_data.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896649       -73.844846   
2                         Carvel Ice Cream       40.890487       -73.848568   
3  Cooler Runnings Jamaican Restaurant Inc       40.898276       -73.850381   
4                                    Shell       40.894187       -73.845862   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2        Ice Cream Shop  
3  Caribbean Restaurant  
4           Gas Station

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

#### Let's find out how many unique categories can be curated from all the returned venues

In [62]:
print("There are {} Unique Category".format(len(neighborhoods_data['Venue Category'].unique())))

There are 427 Unique Category


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [68]:
# one hot encoding
neighbourhood_onehot = pd.get_dummies(neighborhoods_data[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
neighbourhood_onehot['Neighborhood'] = neighborhoods_data['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [neighbourhood_onehot.columns[-1]] + list(neighbourhood_onehot.columns[:-1])
neighbourhood_onehot = neighbourhood_onehot[fixed_columns]

neighbourhood_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  Airport Terminal  Airport Tram  American Restaurant  \
0                   0                 0             0                    0   
1                   0                 0             0                    0   
2                   0                 0             0                    0   
3                   0                 0             0                    0   
4                   0                 0             0                    0   

   Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
0               0             0       0                 0   
1               0             0       0                 0   
2               0             0       0                 0   
3               0             0       0                 0   
4               0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  Auditorium  \
0                     0                 0                   0           0   
1                     0                 0                   0           0   
2                     0                 0                   0           0   
3                     0                 0                   0           0   
4                     0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  Automotive Shop  \
0                      0                    0              0                0   
1                      0                    0              0                0   
2                      0                    0              0                0   
3                      0                    0              0                0   
4                      0                    0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Bath House  Bathing Area  Beach  \
0                 0                 0           0             0      0   
1                 0                 0           0             0      0   
2                 0                 0           0             0      0   
3                 0                 0           0             0      0   
4                 0                 0           0             0      0   

   Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0          0                0         0            0           0   
1          0                0         0            0           0   
2          0                0         0            0           0   
3          0                0         0            0           0   
4          0                0         0            0           0   

   Big Box Store  Bike Renta

And let's examine the new dataframe size.

In [69]:
neighbourhood_onehot.shape

(20582, 427)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [70]:
neighbourhood_grouped = neighbourhood_onehot.groupby('Neighborhood').mean().reset_index()
neighbourhood_grouped.head(5)

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0       Allerton          0.0                0.0             0.0   
1       Annadale          0.0                0.0             0.0   
2  Arden Heights          0.0                0.0             0.0   
3      Arlington          0.0                0.0             0.0   
4       Arrochar          0.0                0.0             0.0   

   Afghan Restaurant  African Restaurant  Airport Terminal  Airport Tram  \
0                0.0                 0.0               0.0           0.0   
1                0.0                 0.0               0.0           0.0   
2                0.0                 0.0               0.0           0.0   
3                0.0                 0.0               0.0           0.0   
4                0.0                 0.0               0.0           0.0   

   American Restaurant  Animal Shelter  Antique Shop  Arcade  \
0             0.000000             0.0           0.0     0.0   
1             0.090909             0.0           0.0     0.0   
2             0.000000             0.0           0.0     0.0   
3             0.000000             0.0           0.0     0.0   
4             0.000000             0.0           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0               0.0                     0.0          0.0         0.0   
1               0.0                     0.0          0.0         0.0   
2               0.0                     0.0          0.0         0.0   
3               0.0                     0.0          0.0         0.0   
4               0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0               0.0   
1                  0.0                   0.0               0.0   
2                  0.0                   0.0               0.0   
3                  0.0                   0.0               0.0   
4                  0.0                   0.0               0.0   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0            0.000000         0.0                    0.0                  0.0   
1            0.000000         0.0                    0.0                  0.0   
2            0.000000         0.0                    0.0                  0.0   
3            0.000000         0.0                    0.0                  0.0   
4            0.058824         0.0                    0.0                  0.0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  \
0            0.0              0.0        0.0         0.0    0.000000   
1            0.0              0.0        0.0         0.0    0.000000   
2            0.0              0.0        0.0         0.0    0.000000   
3            0.0              0.0        0.0         0.0    0.000000   
4            0.0              0.0        0.0         0.0    0.058824   

     Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0  0.000000   0.0  0.0             0.0               0.0               0.0   
1  0.090909   0.0  0.0             0.0               0.0               0.0   
2  0.000000   0.0  0.0             0.0               0.0               0.0   
3  0.000000   0.0  0.0             0.0               0.0               0.0   
4  0.000000   0.0  0.0             0.0               0.0               0.0   

   Bath House  Bathing Area  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
0         0.0           0.0    0.0        0.0              0.0       0.0   
1         0.0           0.0    0.0        0.0              0.0       0.0   
2         0.0           0.0    0.0        0.0              0.0       0.0   
3         0.0           0.0    0.0        0.0              0.0       0.0   
4         0.0           0.0    0.0        0.0              0.0       0.0   

   Beer Garden  Beer Store  Big Box Store  Bike Rental / Bike Share  \
0     

#### Let's confirm the new size

In [71]:
neighbourhood_onehot.shape

(20582, 427)

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [72]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [81]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = neighbourhood_onehot['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(5)

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0    Wakefield           Pizza Place    Italian Restaurant   
1    Wakefield         Women's Store         Event Service   
2    Wakefield                   NaN                   NaN   
3    Wakefield                   NaN                   NaN   
4    Wakefield                   NaN                   NaN   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Deli / Bodega           Coffee Shop    Chinese Restaurant   
1               Exhibit            Eye Doctor               Factory   
2                   NaN                   NaN                   NaN   
3                   NaN                   NaN                   NaN   
4                   NaN                   NaN                   NaN   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Bakery                   Bar        Sandwich Place   
1    Falafel Restaurant                  Farm        Farmers Market   
2                   NaN                   NaN                   NaN   
3                   NaN                   NaN                   NaN   
4                   NaN                   NaN                   NaN   

  9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant               Pharmacy  
1  Fast Food Restaurant                  Field  
2                   NaN                    NaN  
3                   NaN                    NaN  
4                   NaN                    NaN

In [82]:
ny_bar=neighborhoods_venues_sorted[neighborhoods_venues_sorted['1st Most Common Venue']=='Pizza Place']
ny_bar.head(5)

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0    Wakefield           Pizza Place    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Deli / Bodega           Coffee Shop    Chinese Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Bakery                   Bar        Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant               Pharmacy

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [83]:
# set number of clusters
kclusters = 5

neighbourhood_grouped_clustering = neighbourhood_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(neighbourhood_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 
manhattan_grouped_clustering.head(1)

Accessories Store  Adult Boutique  Afghan Restaurant  African Restaurant  \
0                0.0             0.0                0.0                 0.0   

   American Restaurant  Animal Shelter  Antique Shop  Arcade  \
0                 0.01             0.0           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0               0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                  0.0               0.0                0.01        0.01   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                    0.0                  0.0            0.0       0.02   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0         0.0         0.0    0.01   0.0  0.0             0.0   

   Basketball Court  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0               0.0       0.0         0.01         0.0            0.0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                       0.0        0.0         0.0    0.01         0.0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0           0.01        0.0       0.0         0.0                   0.0   

   Breakfast Spot  Bridal Shop  Bubble Tea Shop  Building  Burger Joint  \
0             0.0          0.0              0.0      0.01          0.02   

   Burrito Place  Bus Line  Bus Station  Bus Stop  Business Service  Butcher  \
0           0.01       0.0          0.0       0.0               0.0      0.0   

   Cafeteria  Café  Cajun / Creole Restaurant  Cambodian Restaurant  \
0        0.0   0.0                        0.0                   0.0   

   Camera Store  Candy Store  Cantonese Restaurant  Caribbean Restaurant  \
0           0.0         0.01                   0.0                   0.0   

   Caucasian Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                   0.0          0.0                0.01             0.0   

   Circus  Climbing Gym  Clothing Store  Club House  Cocktail Bar  \
0     0.0           0.0            0.03         0.0           0.0   

   Coffee Shop  College Academic Building  College Bookstore  \
0         0.07                        0.0                0.0   

   College Cafeteria  College Gym  College Theater  Comedy Club  \
0                0.0          0.0              0.0          0.0   

   Community Center  Concert Hall  Convenience Store  Cosmetics Shop  \
0               0.0           0.0                0.0            0.01   

   Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  Cycle Studio  \
0              0.0       0.0               0.0          0.02           0.0   

   Czech Restaurant  Dance Studio  Daycare  Deli / Bodega  Department Store  \
0               0.0           0.0      0.0            0.0              0.02   

   Design Studio  Dessert Shop  Dim Sum Restaurant  Diner  Discount Store  \
0            0.0           0.0                 0.0    0.0             0.0   

   Dive Bar  Dog Run  Donut Shop  Drugstore  Dry Cleaner  Dumpling Restaurant  \
0       0.0      0.0         0.0        0.0          0.0                  0.0   

   Duty-free Shop  Eastern European Restaurant  Electronics Store  \
0             0.0                          0.0               0.01   

   Empanada Restaurant  English Restaurant  Ethiopian Restaurant  Event Space  \
0                  0.0                 0.0                   0.0          0.0   

   Exhibit  Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
0      0.0                 0.0             0.0                   0.0   

   Filipino Restaurant  Fish Market  Flea Market  Flower Shop  \
0                  0.0          0.0          0.0          0.0   

   Food & Drink Shop  Food Court  Food Stand  Food Truck  Fountain  \
0                0.0        0.02         0.0        0.01       0.0   

   French Restaurant  Fried 

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [85]:
# add clustering labels
neighbourhood_grouped_clustering.insert(0, 'Cluster Labels', kmeans.labels_)

neighborhood_merged = neighborhoods_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
neighborhood_merged = neighborhood_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

neighborhood_merged.head() # check the last columns!

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
0    Wakefield              40.894705              -73.847201   
0    Wakefield              40.894705              -73.847201   
0    Wakefield              40.894705              -73.847201   
0    Wakefield              40.894705              -73.847201   

              Venue  Venue Latitude  Venue Longitude Venue Category  \
0  Lollipops Gelato       40.894123       -73.845892   Dessert Shop   
0  Lollipops Gelato       40.894123       -73.845892   Dessert Shop   
0  Lollipops Gelato       40.894123       -73.845892   Dessert Shop   
0  Lollipops Gelato       40.894123       -73.845892   Dessert Shop   
0  Lollipops Gelato       40.894123       -73.845892   Dessert Shop   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Pizza Place    Italian Restaurant         Deli / Bodega   
0         Women's Store         Event Service               Exhibit   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0           Coffee Shop    Chinese Restaurant                Bakery   
0            Eye Doctor               Factory    Falafel Restaurant   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                   Bar        Sandwich Place    Mexican Restaurant   
0                  Farm        Farmers Market  Fast Food Restaurant   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   
0                   NaN                   NaN                   NaN   

  10th Most Common Venue  
0               Pharmacy  
0                  Field  
0                    NaN  
0                    NaN  
0                    NaN

Finally, let's visualize the resulting clusters

In [87]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(neighborhood_merged['Neighborhood Latitude'], neighborhood_merged['Neighborhood Longitude'], neighborhood_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [46]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown    Chinese Restaurant   American Restaurant   
6       Central Harlem    African Restaurant            Public Art   
8      Upper East Side               Exhibit    Italian Restaurant   
13      Lincoln Square  Gym / Fitness Center               Theater   
14             Clinton               Theater  Gym / Fitness Center   
15             Midtown                 Hotel           Coffee Shop   
18   Greenwich Village    Italian Restaurant        Clothing Store   
21             Tribeca                   Spa                  Café   
22        Little Italy                Bakery    Seafood Restaurant   
23                Soho        Clothing Store              Boutique   
24        West Village    Italian Restaurant        Cosmetics Shop   
29  Financial District           Coffee Shop            Steakhouse   
32        Civic Center    Italian Restaurant  Gym / Fitness Center   
33       Midtown South     Korean Restaurant                 Hotel   
35          Turtle Bay                 Hotel    Italian Restaurant   
38            Flatiron                   Gym           Yoga Studio   
39        Hudson Yards   American Restaurant                  Café   
41           Chinatown    Chinese Restaurant   American Restaurant   
46      Central Harlem    African Restaurant            Public Art   
48     Upper East Side               Exhibit    Italian Restaurant   
53      Lincoln Square  Gym / Fitness Center               Theater   
54             Clinton               Theater  Gym / Fitness Center   
55             Midtown                 Hotel           Coffee Shop   
58   Greenwich Village    Italian Restaurant        Clothing Store   
61             Tribeca                   Spa                  Café   
62        Little Italy                Bakery    Seafood Restaurant   
63                Soho        Clothing Store              Boutique   
64        West Village    Italian Restaurant        Cosmetics Shop   
69  Financial District           Coffee Shop            Steakhouse   
72        Civic Center    Italian Restaurant  Gym / Fitness Center   
73       Midtown South     Korean Restaurant                 Hotel   
75          Turtle Bay                 Hotel    Italian Restaurant   
78            Flatiron                   Gym           Yoga Studio   
79        Hudson Yards   American Restaurant                  Café   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1              Cocktail Bar   Dumpling Restaurant       Bubble Tea Shop   
6      Gym / Fitness Center   American Restaurant     French Restaurant   
8               Coffee Shop  Gym / Fitness Center             Juice Bar   
13                     Café          Concert Hall                 Plaza   
14                    Hotel   American Restaurant    Italian Restaurant   
15             Cocktail Bar        Clothing Store               Theater   
18         Sushi Restaurant     French Restaurant                  Café   
21       Italian Restaurant                  Park   American Restaurant   
22                    Hotel        Clothing Store    Salon / Barbershop   
23               Shoe Store         Women's Store    Italian Restaurant   
24  New American Restaurant                  Park              Wine Bar   
29                    Hotel                   Gym             Wine Shop   
32              Coffee Shop        Sandwich Place     French Restaurant   
33                Hotel Bar   Japanese Restaurant           Coffee Shop   
35               Steakhouse      Sushi Restaurant           Coffee Shop   
38     Gym / Fitness Center   American Restaurant   Japanese Restaurant   
39       Italian Restaurant           Coffee Shop               Theater   
41             Cocktail Bar   Dumpling Restaurant       Bubble Tea Shop   
46     Gym / Fitness Center   American Restaurant     French Restaurant   
48              Coffee Shop  Gym / Fitness Center          

#### Cluster 2

In [47]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
9          Yorkville    Italian Restaurant                   Bar   
10        Lenox Hill      Sushi Restaurant    Italian Restaurant   
12   Upper West Side    Italian Restaurant              Wine Bar   
16       Murray Hill                 Hotel   Japanese Restaurant   
17           Chelsea           Coffee Shop        Ice Cream Shop   
19      East Village                   Bar              Wine Bar   
25  Manhattan Valley     Indian Restaurant           Coffee Shop   
27          Gramercy          Cocktail Bar   American Restaurant   
30     Carnegie Hill           Coffee Shop           Pizza Place   
31              Noho    Italian Restaurant     French Restaurant   
34      Sutton Place    Italian Restaurant  Gym / Fitness Center   
49         Yorkville    Italian Restaurant                   Bar   
50        Lenox Hill      Sushi Restaurant    Italian Restaurant   
52   Upper West Side    Italian Restaurant              Wine Bar   
56       Murray Hill                 Hotel   Japanese Restaurant   
57           Chelsea           Coffee Shop        Ice Cream Shop   
59      East Village                   Bar              Wine Bar   
65  Manhattan Valley     Indian Restaurant           Coffee Shop   
67          Gramercy          Cocktail Bar   American Restaurant   
70     Carnegie Hill           Coffee Shop           Pizza Place   
71              Noho    Italian Restaurant     French Restaurant   
74      Sutton Place    Italian Restaurant  Gym / Fitness Center   

   3rd Most Common Venue   4th Most Common Venue     5th Most Common Venue  \
9                    Gym             Coffee Shop          Sushi Restaurant   
10           Coffee Shop             Pizza Place       Sporting Goods Shop   
12                   Bar             Coffee Shop  Mediterranean Restaurant   
16           Coffee Shop          Sandwich Place                       Gym   
17    Italian Restaurant     American Restaurant                 Nightclub   
19    Mexican Restaurant          Ice Cream Shop        Chinese Restaurant   
25           Pizza Place             Yoga Studio        Mexican Restaurant   
27    Italian Restaurant              Bagel Shop                       Bar   
30                  Café             Yoga Studio                 Bookstore   
31          Cocktail Bar      Mexican Restaurant             Grocery Store   
34     Indian Restaurant  Furniture / Home Store       American Restaurant   
49                   Gym             Coffee Shop          Sushi Restaurant   
50           Coffee Shop             Pizza Place       Sporting Goods Shop   
52                   Bar             Coffee Shop  Mediterranean Restaurant   
56           Coffee Shop          Sandwich Place                       Gym   
57    Italian Restaurant     American Restaurant                 Nightclub   
59    Mexican Restaurant          Ice Cream Shop        Chinese Restaurant   
65           Pizza Place             Yoga Studio        Mexican Restaurant   
67    Italian Restaurant              Bagel Shop                       Bar   
70                  Café             Yoga Studio                 Bookstore   
71          Cocktail Bar      Mexican Restaurant             Grocery Store   
74     Indian Restaurant  Furniture / Home Store       American Restaurant   

   6th Most Common Venue   7th Most Common Venue  \
9            Pizza Place          Ice Cream Shop   
10  Gym / Fitness Center            Burger Joint   
12                Bakery       Indian Restaurant   
16     French Restaurant      Salon / Barbershop   
17                Bakery                 Theater   
19          Cocktail Bar             Pizza Place   
25                  Café                     Bar   
27           Pizza Place  Thrift / Vintage Store   
30        Cosmetics Shop       French Restaurant   
31           Art Gallery               Gift Shop   
34           Pizza Place               Juice Bar   
49           Pizza Place

#### Cluster 3

In [48]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                   Bar                  Park   
77  Stuyvesant Town                   Bar                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37            Playground           Pet Service        Farmers Market   
77            Playground           Pet Service        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37        Baseball Field              Fountain       Harbor / Marina   
77        Baseball Field              Fountain       Harbor / Marina   

   9th Most Common Venue 10th Most Common Venue  
37          Cocktail Bar            Coffee Shop  
77          Cocktail Bar            Coffee Shop

#### Cluster 4

In [49]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
11     Roosevelt Island           Coffee Shop                  Park   
26  Morningside Heights                  Park           Coffee Shop   
28    Battery Park City                  Park           Coffee Shop   
51     Roosevelt Island           Coffee Shop                  Park   
66  Morningside Heights                  Park           Coffee Shop   
68    Battery Park City                  Park           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11        Sandwich Place      Greek Restaurant              Bus Stop   
26   American Restaurant             Bookstore            Food Truck   
28                 Hotel         Memorial Site        Clothing Store   
51        Sandwich Place      Greek Restaurant              Bus Stop   
66   American Restaurant             Bookstore            Food Truck   
68                 Hotel         Memorial Site        Clothing Store   

   6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
11        Farmers Market           Supermarket  Outdoors & Recreation   
26          Tennis Court         Deli / Bodega           Burger Joint   
28             Wine Shop    Italian Restaurant                    Gym   
51        Farmers Market           Supermarket  Outdoors & Recreation   
66          Tennis Court         Deli / Bodega           Burger Joint   
68             Wine Shop    Italian Restaurant                    Gym   

   9th Most Common Venue 10th Most Common Venue  
11           Dry Cleaner                 School  
26           Pizza Place         Sandwich Place  
28           Pizza Place             Playground  
51           Dry Cleaner                 School  
66           Pizza Place         Sandwich Place  
68           Pizza Place             Playground

#### Cluster 5

In [69]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          Marble Hill        Discount Store           Coffee Shop   
2   Washington Heights                  Café                Bakery   
3               Inwood                  Café    Mexican Restaurant   
4     Hamilton Heights                  Café           Pizza Place   
5       Manhattanville    Mexican Restaurant         Deli / Bodega   
7          East Harlem         Deli / Bodega    Mexican Restaurant   
20     Lower East Side           Coffee Shop                  Café   
25    Manhattan Valley     Indian Restaurant           Coffee Shop   
36          Tudor City                  Park    Mexican Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Sandwich Place           Yoga Studio        Clothing Store   
2      Mobile Phone Shop    Spanish Restaurant        Sandwich Place   
3            Pizza Place                Lounge                Bakery   
4     Mexican Restaurant         Deli / Bodega           Coffee Shop   
5                   Park    Chinese Restaurant    Seafood Restaurant   
7                 Bakery                   Spa       Thai Restaurant   
20    Chinese Restaurant      Ramen Restaurant           Pizza Place   
25           Pizza Place       Thai Restaurant    Mexican Restaurant   
36      Greek Restaurant                  Café      Asian Restaurant   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                         Gym        Tennis Stadium       Supplement Shop   
2            Tapas Restaurant                  Park         Deli / Bodega   
3          Chinese Restaurant         Deli / Bodega              Wine Bar   
4                        Park  Caribbean Restaurant    Chinese Restaurant   
5          Italian Restaurant           Coffee Shop      Sushi Restaurant   
7   Latin American Restaurant        Clothing Store                  Bank   
20                       Park                Bakery           Art Gallery   
25                Yoga Studio   Szechuan Restaurant                   Bar   
36              Deli / Bodega                 Hotel           Pizza Place   

   9th Most Common Venue   10th Most Common Venue  
0             Steakhouse       Seafood Restaurant  
2     Chinese Restaurant  New American Restaurant  
3                   Park      American Restaurant  
4                 Bakery                   School  
5   Gym / Fitness Center          Other Nightlife  
7             Donut Shop              Coffee Shop  
20        Sandwich Place      Japanese Restaurant  
25         Deli / Bodega                     Café  
36               Dog Run                      Spa

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/) and [Polong Lin](https://www.linkedin.com/in/polonglin/). I hope you found this lab interesting and educational. Feel free to contact us if you have any questions!

This notebook is part of a course on **Coursera** called *Applied Data Science Capstone*. If you accessed this notebook outside the course, you can take this course online by clicking [here](http://cocl.us/DP0701EN_Coursera_Week3_LAB2).

<hr>

Copyright &copy; 2018 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).